### PR TITLE
EC enhancements

### PR DESCRIPTION
**Pull request by jack-fortanix**
Basically a rollup of recent changes to internal version + implicitly declare `handle`/`handle_mut` for all structs via macro

### CODE REVIEW

File: mbedtls/src/pk/mod.rs [hunk 1]
```diff
@@ -139,6 +142,52 @@ impl Pk {
         Ok(ret)
     }
 
+    pub fn generate_ec_using_curve<F: ::rng::Random>(
```
Comment (jethrogb) [inline mbedtls/src/pk/mod.rs]: See comment on the internal PR for the API i'd like to see here

File: mbedtls/src/private.rs [hunk 2]
```diff
@@ -26,6 +26,13 @@ pub fn alloc_vec_repeat<F>(mut f: F, data_at_end: bool) -> ::Result<Vec<u8>>
 where
     F: FnMut(*mut c_uchar, size_t) -> c_int,
 {
+    /*
+    Avoid allocating more than a limited amount of memory. In certain conditions
+    with malformed datastructures, mbedtls may return a too-small error regardless of how much
+    buffer space is provided. This causes a loop which terminates with a out of memory panic.
+    */
+    let max_vector_allocation = 4 * 1024 * 1024;
```
Comment (jethrogb) [inline mbedtls/src/private.rs]: `const` makes more sense I think

File: mbedtls/src/pk/mod.rs (hunk 1 above)
Comment (jack-fortanix) [reply mbedtls/src/pk/mod.rs]: @jethrogb Made those changes but I am running into this
```error[E0277]: `?` couldn't convert the error to `error::Error`
   --> mbedtls/src/pk/mod.rs:134:51
    |
134 |             let curve : EcGroup = curve.try_into()?;
    |                                                   ^ the trait `std::convert::From<impl Into<::Error>>` is not implemented for `error::Error`
    |
    = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
    = help: consider adding a `where error::Error: std::convert::From<impl Into<::Error>>` bound
    = note: required by `std::convert::From::from`
```
and I'm not sure how to fix it.

File: mbedtls/src/lib.rs [hunk 3]
```diff
@@ -7,6 +7,11 @@
  * according to those terms. */
 
 #![deny(warnings)]
+
+#![allow(stable_features)]
+#![feature(universal_impl_trait)]
+#![feature(try_from)]
```
Comment (jethrogb) [inline mbedtls/src/lib.rs]: This is not acceptable. Why is this needed?

Comment (jack-fortanix) [reply mbedtls/src/lib.rs]: Because the old nightly used in core_io build requires a feature for try_from https://travis-ci.com/fortanix/rust-mbedtls/jobs/217291163#L1170 and because
```
#![cfg_attr(feature = "core_io", feature(universal_impl_trait))]
#![cfg_attr(feature = "core_io", feature(try_from))] 
```
doesn't work. Obviously feature doesn't work on stable so yeah wtf how to make this work?

**Comment by jethrogb** [comment]
Can you update the nightly that's used along with the core_io crate?

File: mbedtls/src/lib.rs (hunk 3 above)
Comment (jethrogb) [reply mbedtls/src/lib.rs]: Can you update the nightly that's used along with the core_io crate?

Comment (jack-fortanix) [reply mbedtls/src/lib.rs]: ok

**Comment by Goirad** [comment]
This is the last one needed to match the internal version, plus ed25519

**Comment by jethrogb** [comment]
@Goirad can you please do the review?

**Comment by jack-fortanix** [comment]
@Goirad fyi #44 is needed also, we use that for Ed25519 in backend
